crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,187 @@
+//! Different formatting functions.
+
+use std::time::Duration;
+
+/// Returns value with `n` digits after floating point where `n` is `precision`.
+/// Standard rounding rule applies (if `n+1`th digit >= 5 then round up).
+/// If `value` is +/- infinity or NaN returns `value`.
+pub fn round_with_precision(value: f64, precision: u8) -> f64 {
+    if value.is_infinite()
+        || value.is_nan()
+        // Integers that big can't have fractional part (mantissa is already full).
+        || value.abs() >= (1_i64 << 53) as f64
+        // Binary64 format can only precisely represent up to log_10(2^53) digits.
+        || precision >= 17
+    {
+        return value;
+    }
+    let offset = 10_f64.powi(precision.into());
+    if !(value * offset).is_finite() {
+        return value;
+    }
+    (value * offset).round() / offset
+}
+
+/// Returns number of `(days, hours, minutes, seconds, milliseconds, microseconds)`.
+fn get_duration_parts(duration: &Duration) -> (u16, u8, u8, u8, u16, u16) {
+    // In practice we probably don't need nanoseconds.
+    let micros = duration.as_micros();
+    let (millis, micros) = (micros / 1000, (micros % 1000) as u16);
+    let (sec, millis) = (millis / 1000, (millis % 1000) as u16);
+    let (mins, sec) = (sec / 60, (sec % 60) as u8);
+    let (hours, mins) = (mins / 60, (mins % 60) as u8);
+    let (days, hours) = ((hours / 24) as u16, (hours % 24) as u8);
+    (days, hours, mins, sec, millis, micros)
+}
+
+/// Format string using `days`, `hours`, `minutes`, `seconds`.
+fn format_dhms(days: u16, hours: u8, minutes: u8, seconds: u8) -> String {
+    match (days, hours, minutes, seconds) {
+        (0, 0, 0, s) => format!("{s:2} s"),
+        (0, 0, m, s) => format!("{m:2} m {s:2} s"),
+        (0, h, m, s) => format!("{h:2} h {m:2} m {s:2} s"),
+        (d, h, m, s) => format!("{d:3} d {h:2} h {m:2} m {s:2} s"),
+    }
+}
+
+/// Format string starting with number of days and going bigger from there.
+pub fn time_starting_with_seconds(duration: &Duration) -> String {
+    let (days, hours, minutes, seconds, _, _) = get_duration_parts(duration);
+    format_dhms(days, hours, minutes, seconds)
+}
+
+/// Format string starting with number of milliseconds and going bigger from there.
+/// `precision` is how many digits of microseconds from floating point to the right
+/// will be preserved. Note that this function will always remove all trailing zeros
+/// for microseconds.
```

Comment:
The comment itself _is_ 80 chars long. Why does the `/// ` must be included? The Rust code is generally verbose, and various guidelines say to aim for 80 or 100. So if code in a file takes more horizontal space than 80 (plus the first statement), what this change is for? There is nothing about this (80) in the https://github.com/typst/typst/blob/489a6d54c193075d765606dafd562c27c894b04b/CONTRIBUTING.md either.

Reply:
So... It should be documented in the contributing file?

Reply:
It will shrink documentation that was exactly 80 chars to less, making it more squished and probably less readable. You know that, right? In IDEs that leave the docs as-is, this will be noticeable (e.g., Neovim).

Reply:
Sure, we can add it to the CONTRIBUTING file.

But please adjust it.

Reply:
Under normal circumstances, this would be done by `rustfmt` automatically, but unfortunately the relevant option[^1] is unstable.

[^1]: https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#wrap_comments

Reply:
https://github.com/typst/typst/issues/4797

Reply:
What about this one, then?
https://github.com/typst/typst/pull/4587/files#diff-8b050b218f83192727292deff17a26bde7d134563d7a09d381f92f3a80c3b775R25

It's exactly 80 chars + 4. (Or rather, 4 + 80.)

Reply:
> Under normal circumstances, this would be done by `rustfmt` automatically, but unfortunately the relevant option is unstable, requiring nightly rustc.

Yeah, I remember searching for something like this. Sucks that some formatting features take ages to become stable.

Reply:
> What about this one, then? https://github.com/typst/typst/pull/4587/files#diff-8b050b218f83192727292deff17a26bde7d134563d7a09d381f92f3a80c3b775R25
> 
> It's exactly 80 chars + 4. (Or rather, 4 + 80.)

The usual procedure here is to consider the \`raw\` part as a single block, so you'd send it to the next line.

Reply:
Consider this trick: https://github.com/typst/typst/pull/4587#discussion_r1723619625.

So, a newline before the block and not inside of it?

Reply:
I mean, you're not wrong, but if this wouldn't be an option, which one would you choose?

Reply:
I'd move the whole raw to a newline if possible.